mlrun/track/tracker_manager.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2018 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Type, Union
+
+from mlrun.execution import MLClientCtx
+from mlrun.utils import logger
+
+from .tracker import Tracker
+
+
+class TrackerManager:
+    """
+    A class for handling multiple `Tracker` instances during a run. The manager is a singleton class and should be
+    retrieved using the `mlrun.track.get_trackers_manager` function.
+    """
+
+    def __init__(self, _stale: bool = False):
+        """
+        Initialize a new empty tracker manager.
+
+        :param _stale: An inner attribute to init a trackers manager in a specific staleness state.
+        """
+        self._trackers: List[Type[Tracker]] = []
+        self._stale = _stale
+
+    def add_tracker(self, tracker: Type[Tracker]):
+        """
+        Adds a Tracker to trackers list.
+
+        :param tracker: The tracker class to add
+        """
+        self._trackers.append(tracker)
+        logger.debug("Added tracker", tracker=tracker)
+
+    def pre_run(self, context: MLClientCtx):
+        """
+        Goes over all the managed trackers and calls their `pre_run` method.
+
+        :param context: Active mlrun context.
+        """
+        for tracker in self._trackers:
+            tracker.pre_run(context=context)
+
+    def post_run(self, context: Union[MLClientCtx, dict]) -> Union[MLClientCtx, dict]:
+        """
+        Goes over all the managed trackers and calls their `post_run` method.
+
+        :param context: Active mlrun context.
+
+        :return: The context updated with the trackers products.
+        """
+        # Check if the context received is a dict to initialize it as an `MLClientCtx` object:
+
+        is_context_dict = False
+        if isinstance(context, dict):
+            is_context_dict = True
+            context = MLClientCtx.from_dict(
+                context, include_status=True, store_run=False
+            )
+
+        # Go over the trackers and call the `post_run` method:
+        for tracker in self._trackers:
+            tracker.post_run(context)
+
+        # Commit changes:
+        context.commit()
+
+        # Mark the manager as stale, so it can be re-initialized next run:
+        self._stale = True
```

Comment:
hmm... Setting it as stale will mean that next run will re-import all the packages etc., since it will basically generate a new tracker-manager. This may be ineffective - for example if running a hyper-param run which has many iterations I think this will mean that every iteration will re-initialize the tracker.
I guess this mostly covers a situation where I'm running local runs and want to change configuration between runs, right? Why not initialize the tracker once and just check tracker enablements when doing pre/post runs? This way you configure it once, import packages etc., but you can still turn off trackers per run using configurations.

Reply:
@guy1992l what do you think

Reply:
changed the manager to be singleton, no more stale